### PR TITLE
feat(contributing): add task scheduling

### PR DESCRIPTION
**Pull request by shulanbushangshu**
## Description

This pr adds ''task-scheduling" in contributing 

Related to: https://github.com/autowarefoundation/autoware-documentation/issues/293

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The Reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: docs/contributing/coding-guidelines/ros-nodes/task-scheduling.md [hunk 1]
```diff
@@ -1,5 +1,117 @@
...
+   ```bash
+       git clone https://github.com/tier4/caret.git ros2_caret_ws
+       cd ros2_caret_ws
+   ```
+
+2. Create the src directory and clone repositories into it.
+
+   ```bash
+       mkdir src
+       vcs import src < caret.repos --recursive
```
Comment (takam5f2) [inline docs/contributing/coding-guidelines/ros-nodes/task-scheduling.md:62]: `--recursive` is not necessary.

```suggestion
       vcs import src < caret.repos
```

File: docs/contributing/coding-guidelines/ros-nodes/task-scheduling.md [hunk 2]
```diff
@@ -1,5 +1,117 @@
...
+3. Run setup_caret.sh.
+
+   ```bash
+       ./setup_caret.sh
+   ```
+
+4. Build the workspace.
+
+   ```bash
+       source /opt/ros/galactic/setup.bash(or source /opt/ros/humble/setup.bash)
```
Comment (takam5f2) [inline docs/contributing/coding-guidelines/ros-nodes/task-scheduling.md:74]: Installation of caret for galactic should not be mentioned here, I think.

```suggestion
       source /opt/ros/humble/setup.bash
```

**Review by takam5f2 — COMMENTED** [review]
@shulanbushangshu  
Thank you for your contribution.

To my understanding, Autoware is not implemented without considering task scheduling so far. Autoware relies on task scheduling which is managed by ROS 2 and Linux. Task scheduling is one of the issue to be tackled in the future.

It is better, I think, that this section honestly tells that task scheduling is not designed and Autoware runs on a task scheduling by combination of ROS 2 and Linux kernel. "Introducing task optimized task scheduling mechanism and applying it to Autoware is one of our future work" is required statement here.

**Comment by sykwer** [comment]
@shulanbushangshu

Thank you for writing the first draft for the task scheduling document!

We TIERⅣ members are now working on scheduling issues on Linux platforms. Our opinion is that the task scheduling document should be about how to configure the settings of a kernel scheduler to run Autoware in a predictable way. Based on workload analysis and detailed analysis of the implemenation of the Linux kernel, we plan to publish a specialized functionality for the task scheduling of Autoware on Linux and the documentation about it.

Additionally, we need to deal with double scheduling problem (ROS2 executer and kernel scheduler) and scheduler settings for specialized operating systems. We are also struggling with these problem. Use-defined and centralized task scheduler will be developed specialized for Autoware.

As a result, we’d like to keep this document blank until the functionality like explained above become fixed.
Thanks!

File: docs/contributing/coding-guidelines/ros-nodes/task-scheduling.md [hunk 3]
```diff
@@ -1,5 +1,117 @@
 # Task scheduling
 
-!!! warning
+## Scheduing in autoware
 
-    Under Construction
+The software of autoware is the system of multiple nodes.In ROS2 the executor uses one or more threads of the underlying operating system to invoke the callbacks. The different type of callback has the different priority.  
+In autoware,there are two types about "publish-subscribe":
```
Comment (xmfcx) [inline docs/contributing/coding-guidelines/ros-nodes/task-scheduling.md:6]:
```suggestion
Autoware is a system made of multiple ROS2 nodes.
In ROS2 the executor uses one or more threads of the underlying operating system to invoke the callbacks.
Callbacks are units of work like subscription callbacks, timer callbacks, service calls, and received client responses.
The different types of callback has the different priority. Here are the priorities for:
```
Reference documentation: https://docs.ros2.org/latest/api/rclpy/api/execution_and_callbacks.html#module-rclpy.executors

Also made some small fixes like spaces after commas and periods.

**Comment by xmfcx** [comment]
@sykwer 
> As a result, we’d like to keep this document blank until the functionality as explained above become fixed. Thanks!

I think it's ok to merge this proposal (after reviewing) and when you have a better proposal, it can be improved with that? Is that ok with you?

**Comment by xmfcx** [comment]
@shulanbushangshu 

It's a great documentation addition overall, thanks for the work.

Before it's merged, could you also update the punctuation please? Mainly put a space after commas and periods.

**Comment by sykwer** [comment]
I agree with this opinion.

https://github.com/autowarefoundation/autoware-documentation/pull/282#pullrequestreview-1223911853


> It is better, I think, that this section honestly tells that task scheduling is not designed and Autoware runs on a task scheduling by combination of ROS 2 and Linux kernel.

This opinion most accurately describes reality.

**Comment by xmfcx** [comment]
@sykwer 
> I agree with this opinion.

What do you agree with exactly? What do you think we should do in this PR?

**Comment by shulanbushangshu** [comment]
Ok, thank  for your check. I will continue to pay  attention to  this section.

**Comment by xmfcx** [comment]
Created the issue to address this problem once the solution @sykwer and @takam5f2 suggested is implemented. Still, thank you @shulanbushangshu for working on this documentation!

**Pull request closed without merging**